workspace/resource_repo.go

```diff
@@ -0,0 +1,203 @@
+package workspace
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/databrickslabs/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// ReposAPI exposes the Repos API
+type ReposAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+// NewReposAPI creates ReposAPI instance from provider meta
+func NewReposAPI(ctx context.Context, m interface{}) ReposAPI {
+	return ReposAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// ReposInformation provides information about given repository
+type ReposInformation struct {
+	ID           int64  `json:"id"`
+	Url          string `json:"url"`
+	Provider     string `json:"provider,omitempty" tf:"computed,alias:git_provider"`
+	Path         string `json:"path,omitempty" tf:"computed"`
+	Branch       string `json:"branch,omitempty" tf:"computed"`
+	HeadCommitID string `json:"head_commit_id,omitempty" tf:"computed,alias:commit_hash"`
+}
+
+// ID returns job id as string
+func (r ReposInformation) RepoID() string {
+	return fmt.Sprintf("%d", r.ID)
+}
+
+type createRequest struct {
+	Url      string `json:"url"`
+	Provider string `json:"provider"`
+	Path     string `json:"path,omitempty"`
+}
+
+func (a ReposAPI) Create(r createRequest) (ReposInformation, error) {
+	var resp ReposInformation
+	if r.Provider == "" { // trying to infer Git Provider from the URL
+		r.Provider = getProviderFromUrl(r.Url)
+	}
+	if r.Provider == "" {
+		return resp, fmt.Errorf("git_provider isn't specified and we can't detect provider from URL")
+	}
+
+	err := a.client.Post(a.context, "/repos", r, &resp)
+	return resp, err
+}
+
+func (a ReposAPI) Delete(id string) error {
+	return a.client.Delete(a.context, fmt.Sprintf("/repos/%s", id), nil)
+}
+
+func (a ReposAPI) Update(id string, r map[string]string) error {
+	if len(r) == 0 {
+		return nil
+	}
+	// TODO: update may change ONE OF (url AND provider (optional)), (path), or (branch OR tag).
+	// for URL/provider force re-create as there are limits on what could be done for changing URL/provider
+	if path, ok := r["path"]; ok {
```

Comment:
i think the code would be cleaner if this method accepts this struct

```
type updateRequest struct {
  Branch       string `json:"branch,omitempty"`
  Tag.            string `json:"tag,omitempty"`
}
```

because it's going to be easier to use `common.DataToStructPointer` & `common.StructToData` and there'll be less errors to handle. 

and then the schema might be composed of calling StructToSchema twice?.. usage of reflection simplifies resource maintenance, because we have to deal just with structs...

Reply:
Addition of path, and necessity of handling it differently makes things harder - setting branch/tag here is different from Update method. Although maybe we’ll need th handle it later, when it’s implemented

Reply:
Update of path-only will come in the near future, so it will be easier to change code with current approach